### ISSUE TITLE
Programs refusing to create new files

### ISSUE THREAD

**Issue by anders150** [vmware/open-vm-tools#295]
Hi,
some programs refuse to create new files in a shared folder because they believe there is no space available on the share, for instance Spyder (Python) and maybe Signal Desktop (Instant Messaging).

I have debugged this and found that they seem to expect f_files, f_ffree, and f_favail to show free space. Things start working if I set them to "just something" and rebuild, for instance like this in the file vmhgfs-fuse/filesystem.c:

    252          stat->f_blocks = (totalBytes + HGFS_BLOCKSIZE - 1) / HGFS_BLOCKSIZE;
    253          stat->f_bfree = (freeBytes + HGFS_BLOCKSIZE - 1) / HGFS_BLOCKSIZE;
    254          stat->f_bavail = stat->f_bfree;
    255          stat->f_files = 10;
    256          stat->f_ffree = 10;
    257          stat->f_favail = 10;
    258          stat->f_namemax = 255;

Looking in the file modules/solaris/vmhgfs/filesystem.c, it seems that maybe they could be permanently set to "just something":

    614    #define HGFS_FILES_TOTAL     0x00ffffff
    615    #define HGFS_FILES_FREE      0x00ffefff
    616    #define HGFS_FILES_AVAIL     0x00ffef00

This is my first post, and I would like to ask what I should do? Post this like I do now? Try to create a pull request? Something else?

**Comment by lousybrit** [comment]
Hi there,

Thanks for reporting this.
Can you give me some more context on this?
You seems to indicate that you are running Solaris, is that correct?
If so, what is the version?
If not what OS and version of Linux are you running in the VM?
What is the OS and version of the host you are running the VM?
What is the version of Workstation do you have installed?
Can you tell me what share or shares you have created and where do they access on the host?
What is the actual disk usage of the disk that the shares are using on the host?
(I would assume that they have some reasonable amount of free space.)


I will look at the Solaris code to see what is being done there and see if I can reproduce your issue regardless with my base assumptions to start with.

Steve

**Comment by anders150** [comment]
Hi, sorry for giving incomplete context. Here are the version numbers:
VMware Workstation 14.1.2
Ubuntu 18.04.1 LTS as both guest and host
open-vm-tools-desktop 2:10.2.0-3ubuntu3
I have created a folder ~/Desktop/GuestFolder and mounted it as /mnt/hgfs/host
Disk usage is 100+ GB of free space on the host.

**Comment by lousybrit** [comment]
I just looked at the Solaris code is horribly out of  date, But that is because it is really given such a low priority compared with Linux and  Windows VM support.
Ok, so you are running Ubuntu 18.04 guest OS which makes the Solaris code irrelevant  then. So you found it just to see what it did in comparison.

Now I am confused as you talk about setting and rebuilding ? What are you rebuilding and using?
Why not use the vmhgfs-fuse file system that comes with open-vm-tools 10.2.0 version? Or is that an issue somehow?

Thanks
Steve

**Comment by anders150** [comment]
Sorry for confusing, what I did was install open-vm-tools using apt-get from Ubuntu repositories, and got version 10.2.0. Then these programs I wanted to use didn't work, so I started looking for a reason.

I found that if I changed those lines in filesystem.c (shown above) and recompiled vmhgfs-fuse, and put the new one in /usr/local/bin/, then things started working.

And yes, absolutely, I just looked at the Solaris code for comparison.

**Comment by lousybrit** [comment]
So it seems that the issue you are running into is that the shipping open-vm-tools vmhgfs-fuse version is not returning the correct information for disk space. Is that correct?
And you just have the one shared folder for your VM that you are uising?

Thanks
Steve

**Comment by anders150** [comment]
Yes, correct. Do you think it will be possible to change those lines outlined above, in a future release?
(I have only one shared folder on the VM in question, but the problem and possible solution is the same if I add more folders.)

**Comment by lousybrit** [comment]
I just ran statvfs test application on both of my shares on my Ubuntu 18.04 VM
statvfs /mnt/hgfs/share1
statvfs /mnt/hgfs/share2
where share1 and share2 reside on different disks on the host. Both returned the correct disk sizes and free space.

This was running the latest builds of VMware tools, but the vmhgfs-fuse should not be much different.

I am now going to revert back to the open-vm-tools 10.2.0 package from Ubuntu and run the test again.

Steve

**Comment by lousybrit** [comment]
I am running kernel 4.15.0-34-generic and it gave me tools 10.3.0.5330
and the same test as I mentioned above yielded the same correct results for disk size and free space.

**Comment by lousybrit** [comment]
I installed the 10.2.0 version and the same tests succeeded in giving me the correct disk sizes and space.

So I must be doing something different here. 
what path are you seeing or using when you see incorrect information returned?

Thanks
Steve

**Comment by anders150** [comment]
Hi, thank you for helping! Here are the values that statvfs returns for me, with the problem being the zero values for "f_files" etc.

f_bsize=1024, f_frsize=1024, f_blocks=959200352, f_bfree=495326576, f_bavail=495326576, f_files=0, f_ffree=0, f_favail=0, f_flag=4102, f_namemax=0

**Comment by anders150** [comment]
Hi again, the path I am using when I get incomplete information is /mnt/hgfs/host
Could you send me your output from statvfs? Is it all non-zero?

**Comment by lousybrit** [comment]
Ok, I totally goofed and misread this.
I was somehow looking at the wrong f_bXXXX fields which come back correctly.
As you clearly stated though, the f_files, f_ffree and f_avail for the inodes are all zero in our case.

I will file a bug about this and see if I can get a fix done.
The interesting cases are when the host is non-Linux, which I might look into separately.

Thanks for this.
Steve

**Comment by anders150** [comment]
OK, thank you!
